src/adr/0034-distributed-listeners-configuration.md

```diff
@@ -0,0 +1,106 @@
+# 34. Distributed Mailbox Listeners Configuration
+
+Date: 2020-04-23
+
+## Status
+
+Proposed
+
+Supercedes [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md)
+
+## Context
+
+James enables a user to register additional mailbox listeners.
+
+The distributed James server is handling mailbox event processing (mailboxListener execution) using a RabbitMQ work-queue
+per listener.
+
+Currently, mailbox listeners are determined by the guice bindings of the server and additional mailbox listeners defined
+via configuration files.
+
+While the configuration might be specific for each James server, what actually is defined in RabbitMQ is common. 
+Heterogeneous configuration might then result in unpredictable RabbitMQ resource status. This was left as a limitation
+of [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md).
+
+## Decision
+
+We need to centralize the definition of mailbox listeners.
+
+An event sourcing system will track the configured mailbox listeners.
+
+It will have the following commands:
+
+ - **AddListener**: Add a given listener. This should be rejected if the group is already used.
+ - **RemoveListener**: Remove a given listener.
+
+It will have the following events:
+
+ - **ListenerAdded**: A mailbox listener is added
+ - **ListenerRemoved**: A mailbox listener is removed
+
+A subscriber will react to these events to modify the RabbitMQ resource accordingly by adding queues, adding or removing
+bindings.
+
+This event sourcing system differs from the one defined in
+[26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md) by the
+fact that we should also keep track of listener configuration.
+
+Upon start, James will ensure the **configured mailbox listener event sourcing system** contains the guice injected 
```

Comment:
By doing that, you may have the same herogeneous problem when doing a rolling upgrade of your cluster with a version changing the built-in list of listener, don't you?

Reply:
**case 1 **: James Server version 1 do not require listener A, James server version 2 requires listener A.

Since listener A is registered, James server version 1 can not be rebooted prior being upgraded. (As listener A cannot be instanciated)

**case 2**: James Server version 1 require listener A, James server version 2 do not requires listener A.

Upgrading to James version 2 means  that listener A is still registered as an additional listener, that needs to be manually unconfigured once the rolling upgrade finished. Whish is acceptable in upgrade instruction. We need to make sure the listeners could still be instanciated (even with empty code) for a transition period. Sounds acceptable to me.

Conclusion: it's safe but you are right that this special case deserves a section in the ADR.